# Homework Introduction to Python

$\textbf{Ubah nama di bawah yaa!}$

- Nama: Hanna Swestikaputri
- IPK / Nilai Rapot:

In [ ]:
import numpy as np

# please import as much as you need

# Normal

## 1. Operasi Variabel

> Misal kita memiliki toko sepatu. Modal membeli 1 kodi sepatu adalah Rp 1.000.000. Setiap sepatu dijual seharga Rp 80.000. Buatlah variabel modal, kuantitas, dan harga_jual sesuai informasi di atas. Hitung profit yang didapatkan jika semua sepatu berhasil terjual, dan masukkan ke variabel profit.

Saran langkah-langkah pengerjaan
1. Convert terlebih dahulu kodi dalam jumlah barang (1 kodi = xyz buah sepatu)
2. Gunakan operasi matematika sederhana untuk menghitung profit menggunakan tiga buah variabel yang sudah ditentukan

In [ ]:
# Given data
modal = 1_000_000  # in Rupiah
kuantitas = 20  # 1 kodi equals 20 items
harga_jual = 80_000  # in Rupiah per item

# Profit calculation
profit = (harga_jual * kuantitas) - modal
print(f"Profit: Rp {profit}")



Profit: Rp 600000


Jadi, dari profit di atas, dapat disimpulkan bahwa profit yang didapat adalah Rp 600,000

## 2. Package Python dan Dictionary

> Kita akan coba mengakses data scientific article dari PLOS corpus
    Menggunakan API request dengan kode berikut :
```
import requests
response = requests.get('http://api.plos.org/search?q=title:"Drosophila" AND body:"RNA"&fl=id,abstract')
```
**perhatikan tanda kutipnya tidak salah ketika meng-copy ke dalam python

Pertanyaannya :
Convert response.content dari request tersebut dari JSON Format ke dalam Dictionary, dan print-kan key di dalamnya apa saja

Saran langkah-langkah pengerjaan
1. Ikuti code dari soal, kemudian ekstrak value dari variable response menggunakan dot (.) content menjadi sebuah variabel baru (dictionary)
2. Gunakan library json, convert hasil content menggunakan json.loads
3. Print keys dari dictionary tersebut


In [ ]:
# your code goes here!
import requests
import json

# Fetching data from API
response = requests.get('http://api.plos.org/search?q=title:"Drosophila" AND body:"RNA"&fl=id,abstract')
content = response.content

# Convert to dictionary
data = json.loads(content)

# Print dictionary keys
print(data.keys())


dict_keys(['response'])


Jadi, dari url link di atas, kami menyimpulkan bahwa memiliki keys utama yaitu response

# Intermediate

## 3. Operasi dalam List

> Buatlah list dengan 10 nama kota di indonesia, kemudian:
1. Urutkan nama kota di dalam list berdasarkan abjad (a)
2. Ambil 4 kota pertama dari list tersebut (yang telah diurutkan) (b)
3. Ambil 3 kota terakhir dari list tersebut (yang telah diurutkan) (c)

Saran langkah-langkah pengerjaan
1. Tidak ada ketentuan nama kota (misal berdasarkan provinsi X atau pulau Y), dibebaskan untuk memilih. Buat nama-nama kota tersebut dalam sebuah list.
2. Gunakan fungsi sorted(X) untuk mengurutkan secara kecil-ke-besar (sesuai abjad), dimana X adalah contoh variabel list pada langkah (a)
3. Gunakan hasil poin (a), urutkan nama kota di dalam list berdasarkan abjad, untuk mengerjakan poin (b) dan (c).

In [9]:
# your code goes here!
# Daftar Kota
kota = ["Jakarta", "Bandung", "Surabaya", "Medan", "Yogyakarta", "Bali", "Makassar", "Semarang", "Palembang", "Malang"]

# a) Urutkan nama kota di dalam list berdasarkan abjad
kota_berurutan = sorted(kota)

# b) Ambil 4 kota pertama dari list tersebut
empat_pertama = kota_berurutan[:4]

# c) Ambil 3 kota terakhir dari list tersebut
tiga_terakhir = kota_berurutan[-3:]

print("Kota Berdasarkan Urutan Abjad:", kota_berurutan)
print("4 Kota Pertama:", empat_pertama)
print("3 Kota Terakhir:", tiga_terakhir)

Kota Berdasarkan Urutan Abjad: ['Bali', 'Bandung', 'Jakarta', 'Makassar', 'Malang', 'Medan', 'Palembang', 'Semarang', 'Surabaya', 'Yogyakarta']
4 Kota Pertama: ['Bali', 'Bandung', 'Jakarta', 'Makassar']
3 Kota Terakhir: ['Semarang', 'Surabaya', 'Yogyakarta']


Jadi, dari operasi list di atas, dapat disimpulkan bahwa mengambil tiga terakhir dengan memberi minus sebelum angka. Kota pertama adalah urutan ke-0.

## 4. Function dalam List

> Team sales memiliki list total transaksi dari 10 merchant pada bulan Juni 2020, yaitu:
```
total_transaksi = [ 19, 35, 23, 25, 10, 15, 33, 45, 21, 30 ]
```
1. Trx minimum
2. Rata-rata
3. Trx maksimal

Saran langkah-langkah pengerjaan
1. Copy variabel total_transaksi yang ada di soal ke notebook kita (atau ketik ulang boleh)
2. Kemudian, hitung menggunakan function operasi dalam list yang telah dipelajari di kelas pada topik list. Print hasilnya setiap poin (a), (b), dan (c)

Jadi, dari operasi list di atas, dapat disimpulkan bahwa jika mengambil 4 pertama kota pertama dihitung 0. jika mengambil dari akhir bisa ditambahkan tanda minus.

In [8]:
# your code goes here!
# Transaksi sales Juni 2010
total_transaksi = [19, 35, 23, 25, 10, 15, 33, 45, 21, 30]

# a) Transaksi Minimum
trx_min = min(total_transaksi)

# b) Rerata Transaksi
trx_avg = sum(total_transaksi) / len(total_transaksi)

# c) Transaksi Maksimum
trx_max = max(total_transaksi)

print(f"Transaksi Minimum: {trx_min}")
print(f"Rerata Transaksi: {trx_avg}")
print(f"Transaksi Maksimum: {trx_max}")

Transaksi Minimum: 10
Rerata Transaksi: 25.6
Transaksi Maksimum: 45


Jadi, dari hasil operasi list di atas, dapat disimpulkan bahwa nilai minimal, max, rata-rata adalah 10 ; 25.6 ; 45 secara berturut-turut

# Advanced

## 5. NumPy Array

> Buatlah array dari 5 bilangan ganjil pertama (dimulai dari 1), Kemudian lakukan perintah berikut :
1. Kali-kan array tersebut dengan 10
2. Bagi-kan array (hasil dari bagian a) dengan 3
3. Kali-kan array (hasil dari bagian b) dengan array baru :
`[-45, 43, 3, 56, 5]`
4. Carilah pada index ke-berapa nilai tertinggi dan terendah dari array (hasil dari bagian c)

Saran langkah-langkah pengerjaan
1. HATI-HATI dalam pengerjaan, karena apabila poin sebelumnya salah, maka selanjutnya akan kebawa salah (misal (b) salah, maka poin (c) sampai (d) akan salah)
2. Untuk array 5 bilangan ganjil pertama boleh manual boleh menggunakan fungsi, dimulai dari 1
3. Ikuti perintah soal, gunakan numpy array, jangan list. Sesuaikan dengan permintaan soal pada setiap butir.



In [ ]:
# your code goes here!
import numpy as np

# a) Buatlah array dari 5 bilangan ganjil pertama kemudian dikalikan 10
array = np.array([1, 3, 5, 7, 9]) * 10

# b) Bagikan array dengan 3
array_divided = array / 3

# c) Kalikan array dengan array yang baru
new_array = np.array([-45, 43, 3, 56, 5])
result_array = array_divided * new_array

# d) Carilah pada index ke-berapa nilai tertinggi dan terendah dari array
max_index = np.argmax(result_array)
min_index = np.argmin(result_array)

print(f"Result Array: {result_array}")
print(f"Max value at index: {max_index}")
print(f"Min value at index: {min_index}")


Result Array: [-150.          430.           50.         1306.66666667  150.        ]
Max value at index: 3
Min value at index: 0


Jadi, dari hasil operasi numpy di atas, dapat disimpulkan bahwa...

# Terimakasih, salam dari Bekasi